File: Исходники/Глава 2. Часть 1/Динамическое программирование1/Динамическое программирование1/Табличный метод/25. Бином Ньютона за линию/C++/Олимпиадный/binomial_theorem_linear.cpp
#include <bits/stdc++.h>
#include <cassert>

using namespace std;

int binomial_theorem_linear_n(int n, int k) {
    if (n <= 0 or k <= 0 or k > n) {
        return -1;
    }
    int a[n + 1];
    for (int i = 0; i <= n; i++) {
        a[i] = 0;
    }
    a[k] = 1;
    for (int i = k + 1; i <= n; i++) {
        a[i] = a[i - 1] * i / (i - k);
    }
    return a[n];
}

int binomial_theorem_linear_k(int n, int k) {
    if (n <= 0 or k <= 0 or k > n) {
        return -1;
    }
    int a[k + 1];
    for (int i = 0; i <= k; i++) {
        a[i] = 0;
    }
    a[0] = 1;
    for (int i = 1; i <= k + 1; i++) {
        a[i] = a[i - 1] * (n - i + 1) / i;
    }
    return a[k];
}

File: Исходники/Глава 2. Часть 1/Динамическое программирование1/Динамическое программирование1/Табличный метод/25. Бином Ньютона за линию/Python/binomial_theorem_linear.py
def binomial_theorem_linear_n(n, k):
    if n <= 0 or k <= 0 or k > n:
        return -1
    arr = [0] * (n + 1)
    arr[k] = 1
    for i in range(k + 1, n + 1):
        arr[i] = arr[i - 1] * i // (i - k)
    return arr[n]


def binomial_theorem_linear_k(n, k):
    if n <= 0 or k <= 0 or k > n:
        return -1
    arr = [0] * (k + 1)
    arr[0] = 1
    for i in range(1, k + 1):
        arr[i] = arr[i - 1] * (n - i + 1) // i
    return arr[k]
